src/notation/internal/addmenucontroller.cpp

```diff
@@ -0,0 +1,232 @@
+//=============================================================================
+//  MuseScore
+//  Music Composition & Notation
+//
+//  Copyright (C) 2021 MuseScore BVBA and others
+//
+//  This program is free software; you can redistribute it and/or modify
+//  it under the terms of the GNU General Public License version 2.
+//
+//  This program is distributed in the hope that it will be useful,
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+//  GNU General Public License for more details.
+//
+//  You should have received a copy of the GNU General Public License
+//  along with this program; if not, write to the Free Software
+//  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
+//=============================================================================
+#include "addmenucontroller.h"
+
+using namespace mu::notation;
+using namespace mu::actions;
+
+void AddMenuController::init()
+{
+    controller()->actionsAvailableChanged().onReceive(this, [this](const std::vector<actions::ActionCode>& actionCodes) {
+        m_actionsReceiveAvailableChanged.send(actionCodes);
+    });
+}
+
+bool AddMenuController::contains(const ActionCode& actionCode) const
+{
+    std::map<ActionCode, AvailableCallback > actions = this->actions();
+    return actions.find(actionCode) != actions.end();
+}
+
+bool AddMenuController::actionAvailable(const ActionCode& actionCode) const
+{
+    std::map<ActionCode, AvailableCallback > actions = this->actions();
+
+    if (actions.find(actionCode) == actions.end()) {
+        return false;
+    }
+
+    return actions[actionCode]();
+}
+
+mu::async::Channel<std::vector<ActionCode> > AddMenuController::actionsAvailableChanged() const
+{
+    return m_actionsReceiveAvailableChanged;
+}
+
+std::map<ActionCode, AddMenuController::AvailableCallback > AddMenuController::actions() const
+{
+    static std::map<ActionCode, AvailableCallback > _actions = {
+        { "note-input", std::bind(&AddMenuController::isNoteInputAvailable, this) },
+        { "note-c", std::bind(&AddMenuController::isNoteAvailable, this, NoteName::C, NoteAddingMode::NextChord) },
+        { "note-d", std::bind(&AddMenuController::isNoteAvailable, this, NoteName::D, NoteAddingMode::NextChord) },
+        { "note-e", std::bind(&AddMenuController::isNoteAvailable, this, NoteName::E, NoteAddingMode::NextChord) },
+        { "note-f", std::bind(&AddMenuController::isNoteAvailable, this, NoteName::F, NoteAddingMode::NextChord) },
+        { "note-g", std::bind(&AddMenuController::isNoteAvailable, this, NoteName::G, NoteAddingMode::NextChord) },
```

Comment:
You can use cycles for similar actions like note-* and interval-*

Reply:
it will be difficult to pass parameters in cycles